### PR TITLE
Bitbucket API V2 Support

### PR DESCRIPTION
**Pull request by AlyIbrahim**
@porcelli 
Adding Bitbucket support and Bitbucket Team Support
I'd consider discussing having 3 different parameters for:

- GitHub Organization
- GitLab Groups
- BitBucketTeams

Since they all are different terminology to the same concept.

### CODE REVIEW

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java [hunk 1]
```diff
@@ -25,13 +26,14 @@
 import org.eclipse.jgit.transport.SshTransport;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.jcraft.jsch.Session;
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: please remove unused code.

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java [hunk 2]
```diff
@@ -1,13 +1,14 @@
 package porcelli.me.git.integration.githook.push;
 
+import static java.util.Comparator.comparing;
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: please remove unused code.

File: src/main/java/porcelli/me/git/integration/githook/push/command/SetupRemote.java [hunk 3]
```diff
@@ -9,6 +9,7 @@
 import org.eclipse.jgit.api.errors.GitAPIException;
 import org.eclipse.jgit.lib.Ref;
 import org.eclipse.jgit.lib.StoredConfig;
+
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/command/SetupRemote.java]: please revert this change

File: src/main/java/porcelli/me/git/integration/githook/push/integration/GitHubIntegration.java [hunk 4]
```diff
@@ -10,6 +10,7 @@
 import org.kohsuke.github.GitHub;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/GitHubIntegration.java]: please revert this change

File: pom.xml [hunk 5]
```diff
@@ -59,13 +75,17 @@
       <groupId>ch.qos.logback</groupId>
       <artifactId>logback-classic</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.glassfish.jersey.media</groupId>
+      <artifactId>jersey-media-json-jackson</artifactId>
```
Comment (porcelli) [inline pom.xml:81]: why this dependency?

File: pom.xml [hunk 6]
```diff
@@ -27,22 +27,38 @@
...
         <version>1.2.3</version>
       </dependency>
+      <dependency>
+        <groupId>com.aliction</groupId>
+        <artifactId>bitbucket-client</artifactId>
+        <version>0.2.1</version>
+      </dependency>
+      <dependency>
+        <groupId>org.glassfish.jersey.media</groupId>
+        <artifactId>jersey-media-json-jackson</artifactId>
```
Comment (porcelli) [inline pom.xml:49]: why this dependency?

File: README.md [hunk 7]
```diff
@@ -24,12 +26,12 @@ Clone this repository and build it locally, for this you need `Git`, `Maven` and
...
 $ cd bc-git-integration-push
 $ mvn clean install
-$ mkdir -p $APP_SERVER_HOME/hooks/ && cp target/git-push-2.1-SNAPSHOT.jar $APP_SERVER_HOME/hooks/
+$ mkdir -p $APP_SERVER_HOME/hooks/ && cp target/git-push-2.2-SNAPSHOT.jar $APP_SERVER_HOME/hooks/
 ```
 
 ## How to setup
 
-The application uses a configuration file `.gitremote` that is located under the user home directory, if you don’t want to start from scratch, you can start running the jar file for the first time after the setup above or just as java -jar git-push-2.1-SNAPSHOT.jar, the application will generate a template configuration file in $HOME/.gitremote that you can follow and modify.
+The application uses a configuration file `.gitremote` that is located under the user home directory, if you don’t want to start from scratch, you can start running the jar file for the first time after the setup above or just as java -jar git-push-2.2-SNAPSHOT.jar, the application will generate a template configuration file in $HOME/.gitremote that you can follow and modify.
```
Comment (porcelli) [inline README.md]: rever the version upgrade changes. I'll do it after the merge myself

File: README.md [hunk 8]
```diff
@@ -24,12 +26,12 @@ Clone this repository and build it locally, for this you need `Git`, `Maven` and
 $ git clone https://github.com/porcelli/bc-git-integration-push.git
 $ cd bc-git-integration-push
 $ mvn clean install
-$ mkdir -p $APP_SERVER_HOME/hooks/ && cp target/git-push-2.1-SNAPSHOT.jar $APP_SERVER_HOME/hooks/
+$ mkdir -p $APP_SERVER_HOME/hooks/ && cp target/git-push-2.2-SNAPSHOT.jar $APP_SERVER_HOME/hooks/
```
Comment (porcelli) [inline README.md]: rever the version upgrade changes. I'll do it after the merge myself

File: README.md [hunk 9]
```diff
@@ -56,18 +59,21 @@ gitlabGroup=Group/subgroup
...
+ - Public Bitbucket url should be the api url ex: api.bitbucket.org and not www.bitbucket.com
 
 ## Enabling and Running
 
 Create a post-commit hook template for Business Central and, finally, start the Business Central with the `org.uberfire.nio.git.hooks` properly set.
 
 ```shell
 $ cd $APP_SERVER_HOME
-$ echo "#\!/bin/bash\njava -jar $APP_SERVER_HOME/hooks/git-push-2.1-SNAPSHOT.jar" > hooks/post-commit
+$ echo "#\!/bin/bash\njava -jar $APP_SERVER_HOME/hooks/git-push-2.2-SNAPSHOT.jar" > hooks/post-commit
```
Comment (porcelli) [inline README.md]: rever the version upgrade changes. I'll do it after the merge myself

File: pom.xml [hunk 10]
```diff
@@ -1,12 +1,13 @@
...
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
 
   <groupId>me.porcelli.git-integration</groupId>
   <artifactId>git-push</artifactId>
-  <version>2.1-SNAPSHOT</version>
+  <version>2.2-SNAPSHOT</version>
```
Comment (porcelli) [inline pom.xml]: rever the version upgrade changes. I'll do it after the merge myself

**Review by porcelli — CHANGES_REQUESTED** [review]

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 2 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: Comparator.comparing is used in line 106, removing it will cause compilation errors..

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 1 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: com.jcraft.jsch.Session is used in lines 126, 128, removing it will cause compilation errors ..

File: pom.xml (hunk 5 above)
Comment (AlyIbrahim) [reply pom.xml:81]: For some reason the dependency is not picked from the main library, bitbucket-client, and without it I am getting Exception :+1:
`org.glassfish.jersey.message.internal.MessageBodyProviderNotFoundException: MessageBodyReader not found for media type=application/json`

File: pom.xml (hunk 6 above)
Comment (AlyIbrahim) [reply pom.xml:49]: For some reason the dependency is not picked from the main library, bitbucket-client, and without it I am getting Exception :
org.glassfish.jersey.message.internal.MessageBodyProviderNotFoundException: MessageBodyReader not found for media type=application/json

File: README.md (hunk 7 above)
Comment (AlyIbrahim) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (AlyIbrahim) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (AlyIbrahim) [reply README.md]: Done

File: pom.xml (hunk 10 above)
Comment (AlyIbrahim) [reply pom.xml]: Done

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 1 above)
Comment (porcelli) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: ohh I see what happened, you formatted the code. please revert the changes of this file and it should be ok

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 2 above)
Comment (porcelli) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: same here: you formatted the code. please revert the changes of this file and it should be ok

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    public BitBucketIntegration(final GitRemoteProperties props) {
+        if (props.getToken().isEmpty()) {
+            LOGGER.error(ERROR_LOGIN);
+            throw new RuntimeException(ERROR_LOGIN);
+        } else {
+            LOGGER.info("Connecting using secured token from properties file");
+            bitbucket = new BitbucketV2API(props.getRemoteGitUrl(), props.getLogin(), props.getToken());
+        }
+        user = bitbucket.UserAPI().getLoggedUser();
+        providerURL = props.getRemoteGitUrl();
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: this property is not used, just delete it

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BitBucketIntegration.class);
+
+    private static final String ERROR_LOGIN = "Connecting using username and password is not supported with Bitbucket Cloud, kindly use token instead.";
+
+    private BitbucketV2API bitbucket;
+    private CredentialsProvider credentialsProvider;
+    private BitbucketTeam team = null;
+    private BitbucketUser user;
+    private String providerURL;
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: this property is not used, just delete it

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            repository = new BitbucketRepository(repoName.toLowerCase(), team, "git", true);
+        } else {
+            repository = new BitbucketRepository(repoName, user, "git", true);
+        }
+        try {
+            createdRepository = bitbucket.RepositoryAPI().createRepository(repository);
+            repoURL = createdRepository.getLinks().getClone_https();
+        } catch (BitbucketCreateRepositoryException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: don't swallow an exception, please log and retrow with RuntimeException if this is the case

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        //        LOGGER.info("Bitbucket has no support for Groups");
+        List<BitbucketTeam> teams = null;
+        try {
+            teams = bitbucket.TeamAPI().getUserTeams(BitbucketRole.ADMIN);
+        } catch (BitbucketGetTeamException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        } catch (BitbucketPageException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: don't swallow an exception, please log and rethrow with RuntimeException if this is the case

**Review by porcelli — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 5 above)
Comment (porcelli) [reply pom.xml:81]: so please <scope>runtime</scope> for this case

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 1 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: Done

File: src/main/java/porcelli/me/git/integration/githook/push/GitHook.java (hunk 2 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/GitHook.java]: Done

File: src/main/java/porcelli/me/git/integration/githook/push/command/SetupRemote.java (hunk 3 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/command/SetupRemote.java]: Reverted with no changes

File: src/main/java/porcelli/me/git/integration/githook/push/integration/GitHubIntegration.java (hunk 4 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/GitHubIntegration.java]: Reverted with no changes

File: pom.xml (hunk 5 above)
Comment (AlyIbrahim) [reply pom.xml:81]: Scope runtime added

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 11 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Yeah, this was used for building url, but I used links later and forget to remove.
Deleted

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 12 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Yeah, this was used for building url, but I used links later and forget to remove.
Deleted

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 13 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Done, Logged and re-throwed.

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 14 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Done, Logged and re-throwed

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+            throw new RuntimeException(ERROR_LOGIN);
+        } else {
+            LOGGER.info("Connecting using secured token from properties file");
+            bitbucket = new BitbucketV2API(props.getRemoteGitUrl(), props.getLogin(), props.getToken());
+        }
+        user = bitbucket.UserAPI().getLoggedUser();
+        credentialsProvider = new UsernamePasswordCredentialsProvider(props.getLogin(), props.getToken());
+        if (!props.getBitbucketTeam().isEmpty()) {
+            team = getTeam(props.getBitbucketTeam());
+        } else {
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: empty else?

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    public BitbucketTeam getTeam(final String groupPath) {
+        List<BitbucketTeam> teams = null;
+        try {
+            teams = bitbucket.TeamAPI().getUserTeams(BitbucketRole.ADMIN);
+        } catch (BitbucketGetTeamException e) {
+            LOGGER.error("An unexpected error occurred.", e);
+            throw new RuntimeException(e);
+        } catch (BitbucketPageException e) {
```
Comment (porcelli) [inline src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: you could just catch Exception or Throwable

**Review by porcelli — CHANGES_REQUESTED** [review]
sorry for the nitpick.

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 15 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Removed

File: src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java (hunk 16 above)
Comment (AlyIbrahim) [reply src/main/java/porcelli/me/git/integration/githook/push/integration/BitBucketIntegration.java]: Multiple exceptions combined to Exception

**Review by porcelli — APPROVED** [review]
